pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package compactor
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+
+	"github.com/go-kit/log/level"
+	"github.com/gorilla/mux"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+
+	"github.com/grafana/dskit/tenant"
+
+	"github.com/grafana/mimir/pkg/storage/bucket"
+)
+
+// CreateBlockUpload handles requests for creating block upload sessions.
+func (c *MultitenantCompactor) CreateBlockUpload(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	blockID := vars["block"]
+	tenantID, _, err := tenant.ExtractTenantIDFromHTTPRequest(r)
+	if err != nil {
+		http.Error(w, "invalid tenant ID", http.StatusBadRequest)
+		return
+	}
+
+	level.Debug(c.logger).Log("msg", "creating block upload session", "user", tenantID, "block_id", blockID)
+	// TODO: Verify that block hasn't already been ingested
+
+	w.WriteHeader(http.StatusOK)
+}
+
+// UploadBlockFile handles requests for uploading block files.
+func (c *MultitenantCompactor) UploadBlockFile(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	blockID := vars["block"]
+	pth, err := url.PathUnescape(vars["path"])
+	if err != nil {
+		http.Error(w, fmt.Sprintf("malformed file path: %q", vars["path"]), http.StatusBadRequest)
+		return
+	}
+
+	tenantID, ctx, err := tenant.ExtractTenantIDFromHTTPRequest(r)
+	if err != nil {
+		http.Error(w, "invalid tenant ID", http.StatusBadRequest)
+		return
+	}
+
+	if path.Base(pth) == "meta.json" {
+		http.Error(w, "meta.json is not allowed", http.StatusBadRequest)
+		return
+	}
+
+	dst := path.Clean(path.Join(blockID, pth))
+	elems := strings.SplitN(dst, "/", 2)
+	if elems[0] != blockID || elems[1] == "" {
```

Comment:
This would panic if `dst` doesn't contain a slash somehow.

```suggestion
	if len(elems) != 2 || elems[0] != blockID || elems[1] == "" {
```

Reply:
Example: https://go.dev/play/p/XJ7WvTAZv08

Reply:
@colega that's a good point point, thanks.